packages/flutter/lib/src/widgets/layout_builder.dart

```diff
@@ -142,41 +130,71 @@ class _LayoutBuilderElement extends RenderObjectElement {
     assert(renderObject == this.renderObject);
   }
 }
 
-class _RenderLayoutBuilder extends RenderBox with RenderObjectWithChildMixin<RenderBox> {
-  _RenderLayoutBuilder({
-    LayoutCallback<BoxConstraints> callback,
-  }) : _callback = callback;
-
-  LayoutCallback<BoxConstraints> get callback => _callback;
-  LayoutCallback<BoxConstraints> _callback;
-  set callback(LayoutCallback<BoxConstraints> value) {
+/// Generic mixin for [RenderObject]s created by [ConstrainedLayoutBuilder].
+///
+/// Provides a callback that should be called at layout time, typically in
+/// [RenderObject.performLayout].
+mixin RenderConstrainedLayoutBuilder<ConstraintType extends Constraints, ChildType extends RenderObject> on RenderObjectWithChildMixin<ChildType> {
+  LayoutCallback<ConstraintType> _callback;
+  /// Change the layout callback.
+  set callback(LayoutCallback<ConstraintType> value) {
     if (value == _callback)
       return;
     _callback = value;
     markNeedsLayout();
   }
 
-  bool _debugThrowIfNotCheckingIntrinsics() {
-    assert(() {
-      if (!RenderObject.debugCheckingIntrinsics) {
-        throw FlutterError(
-          'LayoutBuilder does not support returning intrinsic dimensions.\n'
-          'Calculating the intrinsic dimensions would require running the layout '
-          'callback speculatively, which might mutate the live render object tree.'
-        );
-      }
-      return true;
-    }());
-    return true;
+  /// Invoke the layout callback.
+  void layoutAndBuildChild() {
+    assert(_callback != null);
+    invokeLayoutCallback(_callback);
   }
+}
+
+/// Builds a widget tree that can depend on the parent widget's size.
+///
+/// Similar to the [Builder] widget except that the framework calls the [builder]
+/// function at layout time and provides the parent widget's constraints. This
+/// is useful when the parent constrains the child's size and doesn't depend on
+/// the child's intrinsic size. The [LayoutBuilder]'s final size will match its
+/// child's size.
+///
+/// {@youtube 560 315 https://www.youtube.com/watch?v=IYDVcriKjsw}
+///
+/// If the child should be smaller than the parent, consider wrapping the child
+/// in an [Align] widget. If the child might want to be bigger, consider
+/// wrapping it in a [SingleChildScrollView].
+///
+/// See also:
+///
+///  * [SliverLayoutBuilder], the sliver counterpart of this widget.
+///  * [Builder], which calls a `builder` function at build time.
+///  * [StatefulBuilder], which passes its `builder` function a `setState` callback.
+///  * [CustomSingleChildLayout], which positions its child during layout.
+class LayoutBuilder extends ConstrainedLayoutBuilder<BoxConstraints> {
+  /// Creates a widget that defers its building until layout.
+  ///
+  /// The [builder] argument must not be null.
+  const LayoutBuilder({
+    Key key,
+    LayoutWidgetBuilder builder,
+  }) : super(key: key, builder: builder);
+
+  @override
+  LayoutWidgetBuilder get builder;
```

Comment:
Maybe I am missing something, but how does this override work? It doesn't have a body and the class is not marked abstract?

Reply:
Woah I didn't realize that and it passed the tests.  Turns out this is expected, according to the language spec:

#### section: 16.19:
> Let m be an identifier, o an object, L a library, and C a class which is the
class of o or a superclass thereof.
...
The result of a getter lookup for m in o with respect to L starting in class
C is the result of a getter lookup for m in C with respect to L. The result of a
getter lookup for m in C with respect to L is: If C declares a concrete instance
getter named m that is accessible to L, then that getter declaration is the result
of the getter lookup, and we say that the getter was looked up in C. Otherwise,
if C has a superclass S, the result of the getter lookup is the result of a getter
lookup for m in S with respect to L. Otherwise, we say that the getter lookup
has failed.

#### section 10.4:
> The purpose of an abstract method is to provide a declaration for purposes
such as type checking and reflection. In mixins, it is often useful to introduce
such declarations for methods that the mixin expects will be provided by the
superclass the mixin is applied to.
We wish to detect if one declares a concrete class with abstract members.
However, code like the following should work:
```
class Base {
int get one => 1;
}
abstract class Mix {
int get one;
int get two => one + one;
}
class C extends Base with Mix { }
```
> At run time, the concrete method one declared in Base will be executed, and
no problem should arise. Therefore no error should be raised if a corresponding
concrete member exists in the hierarchy.

Reply:
added implementation.